## Homework 1
Probability Review and Priors

Name: Hyun Suk Ryoo (Max Ryoo)

Computing ID: hr2ee

#### Question 1

(15) You are a data scientist and are choosing between three approaches, A, B,
and C, to a problem. With approach A you will spend a total of 4 days coding
and running an algorithm and it will not produce useful results. With approach
B you will spend a total of 3 days coding and running an algorithm and it will
not produce useful results. With approach C you will spend 1 day coding and
running an algorithm and it will produce useful results. You are equally likely to
choose among unselected options. What is the expected time in days for you to
obtain the results you are looking for, if you continue to select an unselected
option when you do not obtain useful results? What is the variance on this time?

#### Solution 1

For this problem you can think of the the possible choices as shown below.

Starting with A
- $A \rightarrow B \rightarrow C$
- $A \rightarrow C$

Starting with B
- $B \rightarrow A \rightarrow C$
- $B \rightarrow C$

Starting with C
- $C$

We always end at $C$ given the information that the algorithm will produce useful results. We are also given that the approaches $A$, $B$, and $C$ are all equally likely. First lets compute how many days each option (steps) will take

Starting with A
- $A \rightarrow B \rightarrow C = 4 \rightarrow 3 \rightarrow 1 = 8$ days
- $A \rightarrow C = 4 \rightarrow 1 = 5$ days

Starting with B
- $B \rightarrow A \rightarrow C = 3 \rightarrow 4 \rightarrow 1 = 8$ days
- $B \rightarrow C = 3 \rightarrow 1 = 4$ days

Starting with C
- $C \rightarrow 1$ days

Additionally we need to compute the probabilities that each approach will be. We are alos given the informatino taht we will choose among the unselected options. 

Starting with A
- $A \rightarrow B \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} \rightarrow 1 = \frac{1}{6}$
- $A \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} = \frac{1}{6}$

Starting with B
- $B \rightarrow A \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} \rightarrow 1 = \frac{1}{6}$
- $B \rightarrow C =  \frac{1}{3} \rightarrow  \frac{1}{2} =  \frac{1}{6}$

Starting with C
- $C \rightarrow  \frac{1}{3}$

Given all the days and propbabilities we can find the expected number of days with the given probabilites.

$D(A_1)$ = The days for the first option starting with A

$P(A_1)$ = The probability for the first option starting with A

$E(X) = D(A_1) \cdot P(A_1) + D(A_2) \cdot P(A_2) + D(B_1) \cdot P(B_1) + D(B_2) \cdot P(B_2) + D(C_1) \cdot P(C_1)$

$E(X) = 8 \cdot \frac{1}{6} + 5 \cdot \frac{1}{6} + 8 \cdot \frac{1}{6} + 4 \cdot \frac{1}{6} + 1 \cdot \frac{1}{3}$

$E(X) = \frac{8}{6} + \frac{5}{6} + \frac{8}{6} + \frac{4}{6} + \frac{2}{6}$

$E(X) = \frac{27}{6} = 4.5$

In order to find the variance let us map the data into data point values.

Expected Days = 4.5

Paths = [ABC, AC, BAC, BC, C]

Proabilities = [$\frac{1}{6}$, $\frac{1}{6}$, $\frac{1}{6}$, $\frac{1}{6}$,$\frac{1}{3}$]

Actual Days = [8, 5, 8, 4, 1]

We can compute the variance of X to be the following formulation.

$VAR[X] = E[(X-E[X])^2] = \Sigma_{x\in dom(X)} (x-E[x])^2 p(x)$

In [2]:
probs = [1/6, 1/6, 1/6, 1/6, 1/3]
days = [8,5,8,4,1]
expected = 0
for i in range(len(days)):
    expected = probs[i]*days[i] + expected
expected

4.5

In [3]:
probs = [1/6, 1/6, 1/6, 1/6, 1/3]
actual = 8,5,8,4,1
var = 0
for i in range(len(actual)):
    var = var + (actual[i] - expected)**2 * probs[i]
var

8.25

From the computation above and code computation we can see that for this problem the expected time to produce results is **4.5** with a variance of **8.25**

#### Question 2
(15) Suppose if it is sunny or not in Charlottesville depends on the weather of
the last three days. Show how this can be modeled as a Markov chain.

#### Solution 2

The decision for tomorrow's weather condition depends on the last three days. Let us say for the sake of explanation we have the following weather data where S= Sunny and N = Not Sunny.

In [4]:
import random
sampleSize = 600
weather = []
while sampleSize:
    random.seed(sampleSize)
    n = random.randint(1,100)
    if n > 50:
        weather.append("S")
    else:
        weather.append("N")
    sampleSize = sampleSize -1
weather[0:10]

['N', 'S', 'N', 'N', 'S', 'N', 'N', 'N', 'N', 'S']

Since the weather condition is dependent on the last three days the following options are possible for the three day sequence.

- Sunny, Sunny, Sunny
- Sunny, Sunny, Not Sunny
- Sunny, Not Sunny, Sunny
- Sunny, Not Sunny, Not Sunny
- Not Sunny, Sunny, Sunny
- Not Sunny, Sunny, Not Sunny
- Not Sunny, Not Sunny, Sunny
- Not Sunny, Not Sunny, Not Sunny

For each of these possibilites it is possible the next day will be sunny or not sunny.

We can consider making a dictionary value where the KEY is the preceding two days and the value is a tuple where the tuple is (number of occurences, the next day is sunny).

In [5]:
count_dict = {
    "SSS": (0,0),
    "SSN" : (0,0),
    "SNS" : (0,0),
    "SNN" : (0,0),
    "NSS" : (0,0),
    "NSN" : (0,0),
    "NNS" : (0,0),
    "NNN" : (0,0)
}
for i in range(3,len(weather)):
    previous = "".join(weather[i-3:i])
    if previous not in count_dict.keys():
        print(previous)
        print(weather[i])
        break
    success, total = count_dict[previous]
    if weather[i] == "S":
        count_dict[previous] = (success+1,total+1)
    else:
        count_dict[previous] = (success,total+1)

In [6]:
count_dict

{'SSS': (25, 58),
 'SSN': (36, 75),
 'SNS': (36, 75),
 'SNN': (39, 79),
 'NSS': (33, 75),
 'NSN': (39, 79),
 'NNS': (39, 78),
 'NNN': (39, 78)}

In [7]:
import pandas as pd
transition_matrix = pd.DataFrame(count_dict).T
transition_matrix.columns = ["Sunny", "Total"]
transition_matrix["Not Sunny"] = (transition_matrix["Total"] - transition_matrix["Sunny"])/transition_matrix["Total"]
transition_matrix["Sunny"] = transition_matrix["Sunny"]/transition_matrix["Total"]
transition_matrix= transition_matrix.drop(["Total"],axis=1)
transition_matrix

Sunny  Not Sunny
SSS  0.431034   0.568966
SSN  0.480000   0.520000
SNS  0.480000   0.520000
SNN  0.493671   0.506329
NSS  0.440000   0.560000
NSN  0.493671   0.506329
NNS  0.500000   0.500000
NNN  0.500000   0.500000

# ADD MORE DETAIL

#### Question 3

(15) Assume a Gaussian distribution for observations, Xi, i = 1, . . . , N with
unknown mean, M, and known variance 5. Suppose the prior for M is Gaussian
with variance 10. How large a random sample must be taken (i.e., what is the
minimum value for N) to specify an interval having unit length of 1 such that
the probability that M lies in this interval is 0.95?

#### Solution 3

In [35]:
prior_variance = 10
known_variance = 5

We know that the interval of 0.95 is the following

$\mu_x \pm 2\sigma_x$

Given this we know that $4\sigma_x = 1 \rightarrow \sigma_x = \frac{1}{4}$

We also know that $Var[M|x, \sigma^2] = \frac{\sigma_0^2 \sigma^2}{\sigma^2 + N\sigma^2}$. which we can solve for $N$

$\sigma^2 + N\sigma^2 = \frac{\sigma_0^2 \sigma^2}{Var[M|x, \sigma^2]}$

$N\sigma^2 = \frac{\sigma_0^2 \sigma^2}{Var[M|x, \sigma^2]} - \sigma^2 $

$N = \frac{\sigma_0^2 \sigma^2}{Var[M|x, \sigma^2] \cdot \sigma^2} - \sigma^2 \cdot \frac{1}{\sigma^2} $

$N = \frac{10*5}{(\frac{1}{4})^2 \cdot 10} - 5 \cdot \frac{1}{10}$

$N = 5*16 - \frac{1}{2}$

In [36]:
5*16 - 0.5

79.5

In order to satisify that the interval of 0.95 will have a unit length of 1, with the given conditions we will need a sample size of **79.5**

#### Question 4

(15) You have started an online business selling books that are of interest to
your customers. A publisher has just given you a large book with photos from
famous 20th century photographers. You think this book will appeal to people
who have bought art books, history books and coffee table books. In an initial
offering of the new book you collect data on purchases of the new book and
combine these data with data from the past purchases (see ArtHistBooks.csv).

Use Bayesian analysis to give the posterior probabilities for purchases of art
books, history books and coffee table books, as well as, the separate probabilities
for purchases of the new book given each possible combination of prior
purchases of art books, history books and coffee table books. Do this by first
using beta priors with values of the hyperparameters that represent lack of prior
information. Then compute these probabilities again with beta priors that show
strong weighting for low likelihood of a book purchase. Compare your results.

#### Solution 4

In [9]:
books = pd.read_csv("datasets/ArtHistBooks.csv")
books

ArtBooks  HistoryBooks  TableBooks  Purchase
0           0             0           1         0
1           0             1           0         0
2           0             0           0         0
3           1             0           1         0
4           1             1           1         0
..        ...           ...         ...       ...
995         1             1           0         1
996         0             1           0         0
997         1             0           1         0
998         1             1           0         0
999         0             1           0         0

[1000 rows x 4 columns]

#### Question 5
(15) The data set CHDdata.csv contains cases of coronary heart disease (CHD)
and variables associated with the patient’s condition: systolic blood pressure, yearly tobacco use (in kg), low density lipoprotein (ldl),
adiposity, family history (0 or 1), type A personality score (typea), obesity (body
mass index), alcohol use, age, and the diagnosis of CHD (0 or 1).

Perform a Bayesian analysis of these data that finds the posterior marginal
probability distributions for the means for the data of patients with and without
CHD. You should first standard scale (subtract the mean and divide by the
standard deviation) all the numeric variables (remove family history and do not
scale CHD). Then separate the data into two sets, one for patients with CHD
and one for patients without CHD.

Your priors for both groups should assume means of 0 for all variables and a
correlation of 0 between all pairs of variables. You should assume all variances
for the variables are 1. Use a prior alpha equal to one plus the number of
predictor variables. Compute and compare the Bayesian estimates for the
posterior means for each group.

For 5 extra credit points, compute the probability of observing a point at least
as extreme as the posterior mean of patients without coronary heart disease
under the posterior distribution for the patients with coronary heart disease.
Then compute the probability of observing a point at least as extreme as the
posterior mean of patients with coronary heart disease under the posterior
distribution for the patients without coronary heart disease.

#### Solution 5

In [12]:
cor_heart = pd.read_csv("datasets/CHDdata.csv")
cor_heart

sbp  tobacco    ldl  adiposity  famhist  typea  obesity  alcohol  age  \
0    160    12.00   5.73      23.11  Present     49    25.30    97.20   52   
1    144     0.01   4.41      28.61   Absent     55    28.87     2.06   63   
2    118     0.08   3.48      32.28  Present     52    29.14     3.81   46   
3    170     7.50   6.41      38.03  Present     51    31.99    24.26   58   
4    134    13.60   3.50      27.78  Present     60    25.99    57.34   49   
..   ...      ...    ...        ...      ...    ...      ...      ...  ...   
457  214     0.40   5.98      31.72   Absent     64    28.45     0.00   58   
458  182     4.20   4.41      32.10   Absent     52    28.61    18.72   52   
459  108     3.00   1.59      15.23   Absent     40    20.09    26.64   55   
460  118     5.40  11.61      30.79   Absent     64    27.35    23.97   40   
461  132     0.00   4.82      33.41  Present     62    14.70     0.00   46   

     chd  
0      1  
1      1  
2      0  
3      1  
4      1  
..   ...  
457    0  
458    1  
459    0  
460    0  
461    1  

[462 rows x 10 columns]

In [15]:
has_cor = cor_heart[cor_heart["chd"] == 1]
has_cor

sbp  tobacco   ldl  adiposity  famhist  typea  obesity  alcohol  age  chd
0    160    12.00  5.73      23.11  Present     49    25.30    97.20   52    1
1    144     0.01  4.41      28.61   Absent     55    28.87     2.06   63    1
3    170     7.50  6.41      38.03  Present     51    31.99    24.26   58    1
4    134    13.60  3.50      27.78  Present     60    25.99    57.34   49    1
7    114     4.08  4.59      14.60  Present     62    23.11     6.72   58    1
..   ...      ...   ...        ...      ...    ...      ...      ...  ...  ...
453  124     1.60  7.22      39.68  Present     36    31.50     0.00   51    1
454  146     0.64  4.82      28.02   Absent     60    28.11     8.23   39    1
455  128     2.24  2.83      26.48   Absent     48    23.96    47.42   27    1
458  182     4.20  4.41      32.10   Absent     52    28.61    18.72   52    1
461  132     0.00  4.82      33.41  Present     62    14.70     0.00   46    1

[160 rows x 10 columns]

In [17]:
no_cor = cor_heart[cor_heart["chd"] == 0]
no_cor

sbp  tobacco    ldl  adiposity  famhist  typea  obesity  alcohol  age  \
2    118     0.08   3.48      32.28  Present     52    29.14     3.81   46   
5    132     6.20   6.47      36.21  Present     62    30.77    14.14   45   
6    142     4.05   3.38      16.20   Absent     59    20.81     2.62   38   
8    114     0.00   3.83      19.40  Present     49    24.86     2.49   29   
12   118     0.00   1.88      10.05   Absent     59    21.57     0.00   17   
..   ...      ...    ...        ...      ...    ...      ...      ...  ...   
452  154     5.53   3.20      28.81  Present     61    26.15    42.79   42   
456  170     0.40   4.11      42.06  Present     56    33.10     2.06   57   
457  214     0.40   5.98      31.72   Absent     64    28.45     0.00   58   
459  108     3.00   1.59      15.23   Absent     40    20.09    26.64   55   
460  118     5.40  11.61      30.79   Absent     64    27.35    23.97   40   

     chd  
2      0  
5      0  
6      0  
8      0  
12     0  
..   ...  
452    0  
456    0  
457    0  
459    0  
460    0  

[302 rows x 10 columns]

In [25]:
## Standardize All numeric Column
normal_has_cor = has_cor[["sbp", "tobacco", "ldl", "adiposity", "typea", "obesity", "alcohol", "age"]]
normal_has_cor = (normal_has_cor - normal_has_cor.mean())/normal_has_cor.std()
#normal_has_cor["famhist"] = has_cor["famhist"]
#normal_has_cor["chd"] = has_cor["chd"]
normal_has_cor

sbp   tobacco       ldl  adiposity     typea   obesity   alcohol  \
0    0.686834  1.163515  0.108789  -0.709869 -0.536151 -0.301262  2.981339   
1    0.011086 -0.990967 -0.484450   0.069389  0.049406  0.511705 -0.652579   
3    1.109177  0.354910  0.414397   1.404046 -0.340965  1.222197  0.195360   
4   -0.411256  1.451019 -0.893426  -0.048208  0.537371 -0.144134  1.458867   
7   -1.255941 -0.259629 -0.403554  -1.915593  0.732557 -0.799972 -0.474589   
..        ...       ...       ...        ...       ...       ...       ...   
453 -0.833598 -0.705260  0.778430   1.637823 -1.804860  1.110613 -0.731262   
454  0.095555 -0.877763 -0.300187  -0.014204  0.537371  0.338637 -0.416913   
455 -0.664661 -0.590259 -1.194539  -0.232396 -0.633744 -0.606409  1.079968   
458  1.615988 -0.238067 -0.484450   0.563864 -0.243372  0.452497 -0.016243   
461 -0.495724 -0.992764 -0.300187   0.749469  0.732557 -2.715113 -0.731262   

          age  
0    0.160225  
1    1.193174  
3    0.723651  
4   -0.121489  
7    0.723651  
..        ...  
453  0.066320  
454 -1.060534  
455 -2.187388  
458  0.160225  
461 -0.403202  

[160 rows x 8 columns]

In [26]:
## Standardize All numeric Column
normal_no_cor = no_cor[["sbp", "tobacco", "ldl", "adiposity", "typea", "obesity", "alcohol", "age"]]
normal_no_cor = (normal_no_cor - normal_no_cor.mean())/normal_no_cor.std()
#normal_no_cor["famhist"] = no_cor["famhist"]
#normal_no_cor["chd"] = no_cor["chd"]
normal_no_cor

sbp   tobacco       ldl  adiposity     typea   obesity   alcohol  \
2   -0.970826 -0.707272 -0.462050   1.069217 -0.038618  0.831768 -0.515813   
5   -0.192398  0.987035  1.136501   1.574822  1.012063  1.230229 -0.076229   
6    0.363623  0.391813 -0.515513  -0.999515  0.696859 -1.204537 -0.566452   
8   -1.193234 -0.729420 -0.274929  -0.587827 -0.353822 -0.214497 -0.571984   
12  -0.970826 -0.729420 -1.317462  -1.790728  0.696859 -1.018752 -0.677944   
..        ...       ...       ...        ...       ...       ...       ...   
452  1.030847  0.801547 -0.611747   0.622793  0.906995  0.100850  1.142944   
456  1.920480 -0.618681 -0.125231   2.327439  0.381655  1.799808 -0.590282   
457  4.366969 -0.618681  0.874531   0.997172  1.222200  0.663095 -0.677944   
459 -1.526847  0.101123 -1.472505  -1.124308 -1.299435 -1.380545  0.455696   
460 -0.970826  0.765557  3.884512   0.877525  1.222200  0.394195  0.342077   

          age  
2    0.480136  
5    0.412943  
6   -0.057403  
8   -0.662133  
12  -1.468440  
..        ...  
452  0.211366  
456  1.219251  
457  1.286443  
459  1.084866  
460  0.076982  

[302 rows x 8 columns]

In [33]:
import numpy as np
posterior_data_has_cor = []
for i in normal_has_cor.columns:
    # when we use conjugate priors, we only need to use Python like a simple calculator
    xbar = np.average(normal_has_cor[i])
    tau0 = 1000 # chosen hyperparameter 
    ## Pick a high tau0 because we know that the variance should be 0, which means a high tau0
    mu0 = 0  # chosen hyperparameter
    tau = 1 # assumed known
    # "getting" the posterior is simply invoking the formula
    posterior_mean = xbar*(n*tau/(tau0 + n*tau )) + mu0*(tau0/(tau0 + n*tau ))
    posterior_precision = tau0 + n*tau 
    posterior_variance = 1/posterior_precision
    posterior_data_has_cor.append([i, posterior_mean, posterior_precision, posterior_variance])

posterior_has_cor = pd.DataFrame(posterior_data_has_cor)
posterior_has_cor.columns = ['Variable', 'Posterior Mean', 'Posterior Percision', 'Posterior Variance']
posterior_has_cor

Variable  Posterior Mean  Posterior Percision  Posterior Variance
0        sbp   -7.852265e-18                 1018            0.000982
1    tobacco    6.379965e-19                 1018            0.000982
2        ldl   -1.128763e-18                 1018            0.000982
3  adiposity    3.140906e-18                 1018            0.000982
4      typea    1.963066e-18                 1018            0.000982
5    obesity    3.729826e-18                 1018            0.000982
6    alcohol    2.159373e-18                 1018            0.000982
7        age   -4.711359e-18                 1018            0.000982

In [34]:
import numpy as np
posterior_data_no_cor = []
for i in normal_no_cor.columns:
    # when we use conjugate priors, we only need to use Python like a simple calculator
    xbar = np.average(normal_no_cor[i])
    tau0 = 1000 # chosen hyperparameter 
    ## Pick a high tau0 because we know that the variance should be 0, which means a high tau0
    mu0 = 0  # chosen hyperparameter
    tau = 1 # assumed known
    # "getting" the posterior is simply invoking the formula
    posterior_mean = xbar*(n*tau/(tau0 + n*tau )) + mu0*(tau0/(tau0 + n*tau ))
    posterior_precision = tau0 + n*tau 
    posterior_variance = 1/posterior_precision
    posterior_data_no_cor.append([i, posterior_mean, posterior_precision, posterior_variance])

posterior_no_cor = pd.DataFrame(posterior_data_no_cor)
posterior_no_cor.columns = ['Variable', 'Posterior Mean', 'Posterior Percision', 'Posterior Variance']
posterior_no_cor

Variable  Posterior Mean  Posterior Percision  Posterior Variance
0        sbp   -1.456049e-18                 1018            0.000982
1    tobacco   -2.184074e-18                 1018            0.000982
2        ldl    1.352046e-18                 1018            0.000982
3  adiposity   -1.352046e-18                 1018            0.000982
4      typea    6.136207e-18                 1018            0.000982
5    obesity    1.471650e-17                 1018            0.000982
6    alcohol   -2.288077e-18                 1018            0.000982
7        age    3.406115e-18                 1018            0.000982

#### Question 6

(10) For each of the following types of distributions, state the support type
(single or multivariable and discrete or continuous), the formula for the PMF or
PDF, the parameters, the support, the mean, and some typical uses of the
distribution. You may use whatever source(s) you want, including for example
Wikipedia.

#### Solution A 

**Bernoulli Distribution**

- Support Type: Single, Discrete
- Formula for PMF or PDF: PMF = $P(k|p) = p^k (1-p)^{1-k}$
- Parameters:  $\begin{cases}
    p \in [0,1] & \text{probability of success} \\
    q=1-p &
  \end{cases}$
- Support: $k \in \{0,1\}$
- Mean: $p$
- Typical Uses: An example of this distribution is flipping a coin. This distribution is used to model any random variable that has two possible outcomes (success, failure)

Sources:
- https://en.wikipedia.org/wiki/Bernoulli_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution B

**Binomial Distribution**

- Support Type: Single, Discrete
- Formula for PMF or PDF: PMF = ${n \choose k}p^k(1-p)^{n-k}$
- Parameters: $\begin{cases}
    p \in [0,1] & \text{probability of success} \\
    n \in \{1,2,3,...\} & \text{# trials} \\
    q=1-p &
  \end{cases}$
- Support: $k \in \{0,1,...,n\}$
- Mean: $np$
- Typical Uses: The binomial distribution is to model the number of success from multiple independent bernouli trials, such as the number of heads in multiple coin flips. 

Sources:
- https://en.wikipedia.org/wiki/Binomial_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution C

**Poisson Distribution**

- Support Type: Single, Discrete
- Formula for PMF or PDF: PMF = $\frac{\lambda^k e^{-\lambda}}{k!}$
- Parameters: $\lambda \in (0, \infty)$
- Support: $k \in \mathbb{N}_0$ (Natural numbers starting from 0)
- Mean: $\lambda$
- Typical Uses: A famous example for this distribution is the number of accidents on a stretch of road per month. This distribution is used to model the number of occurences over a certain fixed amount of time. However, the conditions must be that the events are independent and that there is a fixed mean rate. 

Sources:
- https://en.wikipedia.org/wiki/Poisson_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution D

**Uniform Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $  f(x|a,b) = \begin{cases}
    \frac{1}{b-a} & \text{if $x\in [a,b]$} \\
    0 & \text{otherwise}
  \end{cases}$
- Parameters: $-\infty < a < b < \infty$
- Support: $x \in [a,b]$
- Mean: $\frac{b-a}{2}$
- Typical Uses: This distribution is used to represent when a random variable has values that are equally possible. For example, spinning a spinner which will come to rest at any angle in $[0, 2\pi]$

Sources:
- https://en.wikipedia.org/wiki/Continuous_uniform_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution E

**Beta Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $\begin{cases}
    f(x|\alpha, \beta) = \frac{x^{\alpha-1}(1-x)^{\beta - 1}}{B(\alpha, \beta)} & \text{where} \\
    B(\alpha, \beta) = \frac{\Gamma(\alpha)\Gamma(\beta)}{\Gamma(\alpha + \beta)}
  \end{cases}$
- Parameters: $\begin{cases}
    \alpha \in (0, \infty) \\
    \beta \in (0, \infty)
  \end{cases}$
- Support: $x \in [0,1]$ (possibly excluding endpoints)
- Mean: $\frac{\alpha}{\alpha + \beta}$
- Typical Uses: This distribution is used as a prior probability distribution in Bayesian methods. It is used to model a random variable on an interval of finite length.

Sources:
- https://en.wikipedia.org/wiki/Beta_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution F

**Gamma Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $f(x|k, \theta) = \frac{1}{\Gamma(k)\theta^k}x^{k-1}e^{-\frac{x}{\theta}}$
- Parameters: $\begin{cases}
    k \in (0, \infty) \\
    \theta \in (0, \infty)
  \end{cases}$
- Support: $x \in (0, \infty)$ (possibly excluding endpoints)
- Mean: $k\theta$
- Typical Uses: A typical example is rainfall accumulated in a reservoir. The Gamma distribution is used to model accumulation over a unit of time. 

Sources:
- https://en.wikipedia.org/wiki/Gamma_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution G

**Gaussian Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $f(x|\mu, \sigma) = \frac{1}{\sigma\sqrt{2\pi}} exp(-\frac{1}{2}(\frac{x - \mu}{\sigma})^2)$
- Parameters: $\begin{cases}
    \mu \in \mathbb{R} & \text{mean} \\
    \sigma > 0 & \text{standard deviation}
  \end{cases}$
- Support: $x \in \mathbb{R}$
- Mean: $\mu$
- Typical Uses: This distribution is the most commenly used for biological and physical measurments. 

Sources: 
- https://en.wikipedia.org/wiki/Normal_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution H

**t Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $f(x|\nu, \mu, \sigma) = \frac{\Gamma\left(\frac{\nu + 1}{2}\right)}
{\sqrt{\nu\pi}\hat{\sigma}^2\Gamma(\frac{\nu}{2})}
(1 + \frac{1}{\nu}\frac{(x-\mu)^2}{\hat{\sigma}^2})^{-\frac{\nu + 1}{2}}
$
- Parameters: $\begin{cases}
    \nu > 0 & \text{(d.o.f)} \\
    \mu \in \mathbb{R} & \text{(location)} \\
    \sigma > 0 & \text{(scale)}
  \end{cases}$
- Support: $x \in \mathbb{R}$
- Mean: $\begin{cases}
    \mu & \text{if } \nu > 0 \\
    \text{else} & undef.
  \end{cases}$
- Typical Uses: This distribution is used when a set of samples is used to assume to have a normal distribution. This distribution is similar to normal distribution, but its shape does differ a slight bit.

Sources:
- https://en.wikipedia.org/wiki/Student%27s_t-distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution I

**Cauchy Distribution**

- Support Type: Single, Continous
- Formula for PMF or PDF: PDF = $\frac{1}{\pi\gamma \left[1 + (\frac{x-x_0}{\gamma})^2\right]}$
- Parameters: $\begin{cases}
    x_0 \in \mathbb{R} & \text{(location)} \\
    \gamma > 0 & \text{(location)} \\
  \end{cases}$
- Support: $x \in (-\infty, \infty)$
- Mean: $undefined$
- Typical Uses: The typical use case for the Cauchy distribution is for the use as the canonical example of a pathological distribution.

Sources:
- https://en.wikipedia.org/wiki/Cauchy_distribution

#### Solution J

**Multinomial Distribution**

- Support Type: Single, Discrete
- Formula for PMF or PDF: PMF = $\frac{n!}{x_1! \cdot \cdot \cdot x_k!}p_1^{x_1} \cdot \cdot \cdot p_k^{x_k}$
- Parameters: $\begin{cases}
    n > 0 & \text{(number of trials)} \\
    k > 0  & \text{(numbeer of mutually exclusive events (integer))} \\
    p_1, ..., p_k & \text{event probabilities) } (\Sigma p_i = 1)
  \end{cases}$
- Support: $\begin{cases}
   x_1 \in \{0,...,n\} \\
    i \in \{1, ..., k\} \\
    \text{with } \Sigma_i x_i=n
  \end{cases}$
- Mean: $E(X_i) = np_i$
- Typical Uses: This distribution is a generalization of the binomial distribution.

Sources:
- https://en.wikipedia.org/wiki/Multinomial_distribution

#### Solution K

**Dirichlet Distribution**

- Support Type: Multivariate, Continous
- Formula for PMF or PDF: PDF = $\begin{cases}
    \frac{1}{b-a} & \text{if $x\in [a,b]$} \\
    0 & \text{otherwise}
  \end{cases}$
- Parameters: $\begin{cases}
    \mu \in \mathbb{R} & \text{(mean)} \\
    \Sigma & \text{(covariance matrix)} \\
  \end{cases}$
- Support: $x \in [a,b]$
- Mean: $\frac{b-a}{2}$
- Typical Uses: This distribution is the multivariate generalizatino of the beta distribution. It is commonly used as prior distributions in Bayesian statistics. Also, the Dirichlet distribution is the conjugate prior of the categorical distribution and multinomial distribution.

Sources:
- https://en.wikipedia.org/wiki/Dirichlet_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution L

**Multivariate Gaussian Distribution**

- Support Type: Multivariate, Continous
- Formula for PMF or PDF: PDF = $\frac{1}{(2\pi)^{-\frac{k}{2}} |\Sigma|^{-\frac{1}{2}}}e^{-\frac{1}{2}(x-\mu)^T\Sigma^{-1}(x-\mu)}$
- Parameters: $\begin{cases}
    \mu \in \mathbb{R}^k & \text{(mean)} \\
    \Sigma & \text{(covariance matrix)} \\
  \end{cases}$
- Support: $x \in \mathbb{R}^k$
- Mean: $\mu$
- Typical Uses: This distribution is the multivariate generalization of the normal distribution. The distribution is used to describe a set of corrleated real-valued random variables that cluster around a mean value. 

Sources:
- https://en.wikipedia.org/wiki/Multivariate_normal_distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution M

**Multivariate t Distribution**

- Support Type: Multivariate, Continous
- Formula for PMF or PDF: PDF = $f(x|\nu, \mu, \Sigma) = \frac{\Gamma(\nu+k)}{\Gamma(\frac{\nu}{2})(\nu\pi)^{\frac{k}{2}}|\Sigma|^{\frac{1}{2}}} [1 + \frac{1}{2}(x-\mu)^T\Sigma^{-1}(x-\mu)]^{-\frac{\nu + k}{2}}$
- Parameters: $\begin{cases}
    \mu \in \mathbb{R}^k & \text{(mean)} \\
    \Sigma & \text{(covariance matrix)} \\
  \end{cases}$
- Support: $x \in \mathbb{R}^k$
- Mean: $\mu$
- Typical Uses: This distribution is the multivariate generalization of the t distribution.

Sources:
- https://en.wikipedia.org/wiki/Multivariate_t-distribution
- https://github.com/wbasener/BayesianML/blob/main/BasenerBrownBML_1_18_2021.pdf

#### Solution N

**Wishart Distribution**

- Support Type: Multivariate, Continous
- Formula for PMF or PDF: PDF = $\begin{cases}
    f_X(x) = \frac{|x|^{\frac{n-p-1}{2}}e^{-\frac{-tr(V^{-1}x)}{2}}}{2^{\frac{np}{2}}|V|^{\frac{n}{2}} \Gamma_p (\frac{n}{2})} \\
    \Gamma_p \text{ is the multivariate gamma function} \\
    tr \text{ is the trace function}
  \end{cases}$
- Parameters: $\begin{cases}
    n > p-1 & \text{(degrees of freedom)} \\
    V > 0 & \text{(scale matrix)} (p \times p \text{ pos. def}) \\
  \end{cases}$
- Support: $X(p \times p)$ positive definite matrix
- Mean: $E[X] = nV$
- Typical Uses: This distribution is the multivariate generalization of the gamma distribution. The wishart distribution is the conjugate prior of the inverse covariance-matrix of a multivariate-normal random-vector

Sources:
- https://en.wikipedia.org/wiki/Wishart_distribution

#### Question 7

(15) Using the Python Notebook https://www.kaggle.com/billbasener/pt2-
probabilities-likelihoods-and-bayes-theorem, complete the challenge question
from Section 6: Modify the code from Section 5 to and add the ability to use the
posterior from conjugate prior function to output the posterior probability
parameters given parameters and for a Gaussian Likelihood with known variance
σ2, and use your modified function to create the Prior, Likelihood, Posterior plots
as in Section 5 of the notebook.

#### Solution 7